### PR TITLE
fix(timeline): correctly filter and display posts in timeline

### PR DESCRIPTION
**Pull request by akhileshthite**
Still not able to view the posts.

If I remove the `timeline` from searchNotes then it's showing but with all the replies
```
  async * searchNotes ({ attributedTo, inReplyTo } = {}, { skip = 0, limit = DEFAULT_LIMIT, sort = -1 } = {}) {
    const tx = this.db.transaction(NOTES_STORE, 'readonly')
    const indexName = inReplyTo ? IN_REPLY_TO_FIELD : (attributedTo ? `${ATTRIBUTED_TO_FIELD}, published` : PUBLISHED_FIELD)
    const index = tx.store.index(indexName)
    const direction = sort > 0 ? 'next' : 'prev'
    let cursor = await index.openCursor(null, direction)

    if (sort === 0) { // Random sort
      const totalNotes = await index.count()
      for (let i = 0; i < limit; i++) {
        const randomSkip = Math.floor(Math.random() * totalNotes)
        cursor = await index.openCursor()
        if (randomSkip > 0) {
          await cursor.advance(randomSkip)
        }
        if (cursor) {
          yield cursor.value
        }
      }
    } else {
      if (attributedTo) {
        cursor = await index.openCursor([attributedTo], direction)
      } else if (inReplyTo) {
        cursor = await index.openCursor(inReplyTo, direction)
      } else {
        cursor = await index.openCursor(null, direction)
      }

      if (skip) await cursor.advance(skip)

      let count = 0
      while (cursor) {
        if (count >= limit) break
        count++
        yield cursor.value
        cursor = await cursor.continue()
      }
    }

    await tx.done
  }
```
Screenshot 2024-08-27 at 11 48 42 AM

### CODE REVIEW

**Comment by akhileshthite** [comment]
> Still not able to view the posts.
> 
> If I remove the `timeline` from searchNotes then it's showing but with all the replies
> 
> ```
>   async * searchNotes ({ attributedTo, inReplyTo } = {}, { skip = 0, limit = DEFAULT_LIMIT, sort = -1 } = {}) {
>     const tx = this.db.transaction(NOTES_STORE, 'readonly')
>     const indexName = inReplyTo ? IN_REPLY_TO_FIELD : (attributedTo ? `${ATTRIBUTED_TO_FIELD}, published` : PUBLISHED_FIELD)
>     const index = tx.store.index(indexName)
>     const direction = sort > 0 ? 'next' : 'prev'
>     let cursor = await index.openCursor(null, direction)
> 
>     if (sort === 0) { // Random sort
>       const totalNotes = await index.count()
>       for (let i = 0; i < limit; i++) {
>         const randomSkip = Math.floor(Math.random() * totalNotes)
>         cursor = await index.openCursor()
>         if (randomSkip > 0) {
>           await cursor.advance(randomSkip)
>         }
>         if (cursor) {
>           yield cursor.value
>         }
>       }
>     } else {
>       if (attributedTo) {
>         cursor = await index.openCursor([attributedTo], direction)
>       } else if (inReplyTo) {
>         cursor = await index.openCursor(inReplyTo, direction)
>       } else {
>         cursor = await index.openCursor(null, direction)
>       }
> 
>       if (skip) await cursor.advance(skip)
> 
>       let count = 0
>       while (cursor) {
>         if (count >= limit) break
>         count++
>         yield cursor.value
>         cursor = await cursor.continue()
>       }
>     }
> 
>     await tx.done
>   }
> ```
> 
> Screenshot 2024-08-27 at 11 48 42 AM

Oh, one more issue; it's showing all the posts with replies for the non-followed actors, is it because we're ingesting the replies for all the actors?

We've fixed this for posts in the past:
https://github.com/hyphacoop/reader.distributed.press/pull/21

**Comment by RangerMauve** [comment]
Tell me the steps you've taken to debug this. THe issue seemed to be with the `following` flag not being set properly in the timeline attribute. Was that not the case?

**Comment by akhileshthite** [comment]
> Tell me the steps you've taken to debug this. THe issue seemed to be with the `following` flag not being set properly in the timeline attribute. Was that not the case?

It is set correctly right?

`ingestActor`:
```
    const isFollowing = await this.isActorFollowed(url)
    if (isFollowing) {
      for await (const note of this.searchNotes({ attributedTo: actor.id })) {
        if (!note.timeline.includes(TIMELINE_FOLLOWING)) {
          note.timeline.push(TIMELINE_FOLLOWING)
          await this.db.put(NOTES_STORE, note)
        }
      }
    }
```

`ingestNote`:
```
    if (isFollowingAuthor && !note.inReplyTo) {
      note.timeline.push(TIMELINE_FOLLOWING)
    }
```

**Comment by RangerMauve** [comment]
Set some breakpoints in there to make sure it's all being set correctly and that the spelling is correct. When I looked in indexedDB I saw the `timeline` field only had `all` set so something is up there.

**Comment by akhileshthite** [comment]
Everything works great, but I noticed that when I view an actor's replies/posts and then return to the timeline, it shows both the latest post and the reply I viewed instead of followed actors's posts.

Screenshot 2024-09-10 at 3 10 53 PM

**Comment by RangerMauve** [comment]
Debugged with akhi. Issue was that the `timelines` field was being indexed as an array and `multiEntry` doesn't support indexing by both the timelines and the published time.

We have two options:

1. We keep the timeline logic as is but instead of using a custom index we can query the `published` index and filter inside searchNotes by skipping the entry instead of yielding it. This means we'd be touching more data from the DB than strictly necessary.

2. We remove the timeline field and instead add new fields to the Notes object for each timeline type and index based on the type. e.g. `note.timeline_following = true` and `createIndex(['timeline_following', 'published'])`. this means we'd need to update more code and custom timelines would be more annoying to add in the future but it will mean our db search is a lot more efficient.

@akhileshthite pick whichever path sounds quickest to get demoable for tomorrow and we can review performance after.

**Review by akhileshthite — COMMENTED** [review]
Working, went with option 1.

File: db.js [hunk 1]
```diff
@@ -253,46 +253,40 @@ export class ActivityPubDB extends EventTarget {
...
-        if (cursor) {
-          yield cursor.value
-        }
-      }
+    let indexName, keyRange
+
+    if (inReplyTo) {
+      indexName = IN_REPLY_TO_FIELD
+      keyRange = IDBKeyRange.only(inReplyTo)
+    } else if (timeline) {
```
Comment (RangerMauve) [inline db.js:263]: Don't use the timeline index at all, just use the published index and filter the timeline when iterating

File: db.js [hunk 2]
```diff
@@ -253,46 +253,40 @@ export class ActivityPubDB extends EventTarget {
...
-      }
+      indexName = PUBLISHED_FIELD
+      keyRange = null
+    }
+
+    const index = tx.store.index(indexName)
+    let cursor = await index.openCursor(keyRange)
+
+    const notes = []
+    while (cursor) {
```
Comment (RangerMauve) [inline db.js:278]: This introduces a performance issue. Instead we should use the published index which is already sorted and filter the field.

File: timeline.js [hunk 3]
```diff
@@ -97,22 +99,36 @@ class ReaderTimeline extends HTMLElement {
...
     // Fetch notes and render them as they become available
+    let notesFound = false
     for await (const note of db.searchNotes({ timeline: 'following' }, { skip: this.skip, limit: this.limit, sort: sortValue })) {
-      this.appendNoteElement(note)
-      count++
+      notesFound = true
+      console.log('Loading note:', note)
+
+      // Exclude replies from appearing in the timeline
+      if (!note.inReplyTo) {
```
Comment (RangerMauve) [inline timeline.js:108]: This should be done inside searchNotes in the timeline filtering logic

**Review by RangerMauve — COMMENTED** [review]
Plz fix these in a new PR

**Pull request merged**